### PR TITLE
Process model

### PR DESCRIPTION
**Pull request by nupplaphil**
This is part of my thoughts about making Friendica workable on different nodes.

I added a whole Model around "Process" which will now be used for each Friendica process.

I moved the `dba::*` calls of the `process`-table as far as possible inside this model and added all of the `proc_open()` and `posix_kill` too. 

It's now possible to add a parent-child relationship between the daemon and the workers so we are now able to check only local process if they are running.

I know that the (new) unittests are still failing (strangely they all succeed in my local environment), but I don't know if I can fix them today but I want your feedback before I go on :-).

So this is **WIP** until the unittests succeed and at least @MrPetovan and @annando approved this PR.

### CODE REVIEW

File: index.php [hunk 1]
```diff
@@ -74,6 +75,9 @@
 
 	Config::init();
 	Session::init();
+	$trace = debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, 1);
+	$command = basename($trace[0]['file']);
+	Process::start($command, Process::PROC_APP);
```
Comment (nupplaphil) [inline index.php]: I'm a unsure with this part.

Is the process of the "Frontend-APP" the php-fpm process or the request of the user? How do you see that?
Will `getmypid()` return the same PID for each request from the same php-fpm process or is each request a unique PID?

File: src/Model/Process.php [hunk 2]
```diff
@@ -5,68 +5,310 @@
...
+		$this->type = $type;
+		$this->code = $this->createCode();
+	}
+
+	/**
+	 * Generates a process code for the logging
+	 *
+	 * @return string a generated process code
+	 */
+	private function createCode()
```
Comment (MrPetovan) [inline src/Model/Process.php]: Can't this be a `__toString()` magic method?

File: index.php (hunk 1 above)
Comment (MrPetovan) [reply index.php]: There can be multiple `php-fpm` processes per pool, so `getmypid()` will return the specific PID of the current `php-fpm` spawn processing the request.

File: src/Model/Process.php (hunk 2 above)
Comment (nupplaphil) [reply src/Model/Process.php]: Because we're extensively use the code for logging, is it better to pre-create the code or is PHP fast enough so that the code will get created on demand?

File: index.php (hunk 1 above)
Comment (nupplaphil) [reply index.php]: Hm.. I think I will remove this part..

My goal was to see the whole System processes in the process table as a kind of monitoring for the admin page later, but I have the fear to make it to complicated

File: bin/daemon.php [hunk 3]
```diff
@@ -62,41 +63,45 @@
...
-if (!empty($pid) && posix_kill($pid, 0)) {
-	die("Daemon process $pid is already running.\n");
+if (Process::isRunning($id)) {
+	die("Daemon process '" . $id . "' is already running.\n");
 }
 
 logger('Starting worker daemon.', LOGGER_DEBUG);
 
+$trace = debug_backtrace(DEBUG_BACKTRACE_IGNORE_ARGS, 1);
+$command = basename($trace[0]['file']);
```
[the remark this answers is not in the document]
Comment (nupplaphil) [reply bin/daemon.php]: I had it from here
https://github.com/friendica/friendica/blob/d3a2ed85fe425cffc103030a6027c9b91f9afc69/src/Core/Worker.php#L1127-L1129
But if `__FILE__` is a easier way to find it, I'll use it :-)

File: src/Database/DBStructure.php [hunk 4]
```diff
@@ -1557,12 +1557,17 @@ public static function definition() {
...
+						"id" => ["type" => "int unsigned", "not null" => "1", "extra" => "auto_increment", "primary" => "1", "comment" => "Global Unique ID"],
+						"pid" => ["type" => "int unsigned", "comment" => "PID of the process"],
+						"parent" => ["type" => "int unsigned", "comment" => "ID of the parent process"],
+						"type" => ["type" => "varchar(5)", "not null" => "1", "comment" => "Type of the process"],
 						"command" => ["type" => "varbinary(32)", "not null" => "1", "default" => "", "comment" => ""],
 						"created" => ["type" => "datetime", "not null" => "1", "default" => NULL_DATE, "comment" => ""],
 						],
 				"indexes" => [
-						"PRIMARY" => ["pid"],
+						"PRIMARY" => ["id"],
```
Comment (annando) [inline src/Database/DBStructure.php]: I don't think that changing the primary index does work with our database update routine.

File: src/Database/DBStructure.php [hunk 5]
```diff
@@ -1557,12 +1557,17 @@ public static function definition() {
...
+						"id_parent" => ["id", "parent"],
+						"id_pid" => ["id", "pid"],
```
Comment (annando) [inline src/Database/DBStructure.php]: What is the sense of these indexes? If "id" is unique, then adding the parent doesn't make sense.

File: bin/daemon.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (MrPetovan) [reply bin/daemon.php]: Without meaning any disrespect for my predecessors, you probably shouldn't trust any code you found elsewhere in this project, 😅

[the remark this answers is not in the document]
Comment (MrPetovan) [reply bin/daemon.php]: Beyond that, this specific trace extraction was meant to work from the method call (looking one stack trace level above), and probably won't work as much from the `bin/daemon.php` file itself. It all depends on what you're expecting to get.

File: index.php (hunk 1 above)
Comment (MrPetovan) [reply index.php]: Why would it be too complicated? If each Friendica process can self-declare itself in the process table, then what's the matter?

File: bin/daemon.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (nupplaphil) [reply bin/daemon.php]: Ah OK, I see. That's why I asked for feedback - especially for PHP specific behaviors :-)

I thought using existing functions may be better rather than refinding the wheel, but I will become more critical to the codebase ;-)

File: src/Model/Process.php (hunk 2 above)
Comment (MrPetovan) [reply src/Model/Process.php]: I'll admit I'm not sure how to make `__toString()` work with the Singleton design pattern in PHP.

File: src/Database/DBStructure.php (hunk 4 above)
Comment (nupplaphil) [reply src/Database/DBStructure.php]: What's the best way to change it? Adding a migration method?

File: src/Database/DBStructure.php (hunk 5 above)
Comment (nupplaphil) [reply src/Database/DBStructure.php]: It should cover the most used accesspath and increase their access speed. So if I select with ID and parent, it's enough to have parent indexed?

File: index.php (hunk 1 above)
Comment (nupplaphil) [reply index.php]: Because the question is how the process will get removed after the call. Hm... But maybe a simple Process::stop at the end should be enough.. I've to test it..

File: src/Database/DBStructure.php (hunk 5 above)
Comment (annando) [reply src/Database/DBStructure.php]: When "id" is unique it doesn't make sense to query for "id" and another field. But I have to have a deeper look at the code to give a better feedback.

File: src/Database/DBStructure.php (hunk 4 above)
Comment (annando) [reply src/Database/DBStructure.php]: Currently there is no way. But possibly we should add some way. I already wanted to add some stuff like this changing of primary fields and removing of fields.

File: src/Database/DBStructure.php (hunk 5 above)
Comment (nupplaphil) [reply src/Database/DBStructure.php]: Business Case:
Give me the process with ID = ? and parent = same parent as the current process
if result => check PID
if not => not same node, so don't check

File: src/Model/Process.php [hunk 6]
```diff
@@ -101,7 +93,7 @@ public static function getCode() {
 			return '';
 		}
 
-		return self::$instance->code;
+		return self::$instance;
```
Comment (MrPetovan) [inline src/Model/Process.php]: Unless the return is type-casted as string, I'm not sure this would have the intended effect. Sure, if it's concatenated in a string, it will work, but it still returns the whole object.

File: src/Database/DBStructure.php (hunk 4 above)
Comment (nupplaphil) [reply src/Database/DBStructure.php]: So it's better to wait for your changes, isn't it? I don't want to include these routines in this PR.
Maybe we can add another PR with exactly this changes and merge it into this PR afterwards?

File: src/Model/Process.php [hunk 7]
```diff
@@ -5,68 +5,302 @@
...
 	public static function deleteInactive()
 	{
+		if (is_null(self::$instance) || is_null(self::$instance->parent)) {
+			return;
+		}
+
 		dba::transaction();
 
-		$processes = dba::select('process', ['pid']);
+		$processes = dba::select('process', ['id', 'pid'], ['parent' => self::$instance->parent ]);
```
Comment (nupplaphil) [inline src/Model/Process.php]: Reminder to me (Because I'm currently at work): parent could be the current instance too => add it to the selection

File: src/Database/DBStructure.php (hunk 4 above)
Comment (annando) [reply src/Database/DBStructure.php]: It's complicated, especially since you want to add some auto-increment field. I have to see what we can do.

File: bin/daemon.php [hunk 8]
```diff
@@ -62,41 +63,45 @@
 	die("Unexpected script behaviour. This message should never occur.\n");
 }
 
-$pid = @file_get_contents($pidfile);
+$id = @file_get_contents($pidfile);
```
Comment (Quix0r) [inline bin/daemon.php]: `$pid` was a more precise name. Can you please revert this back?

File: bin/worker.php [hunk 9]
```diff
@@ -47,16 +48,17 @@
 
 Addon::loadHooks();
 
-$spawn = (($_SERVER["argc"] == 2) && ($_SERVER["argv"][1] == "spawn"));
+$parentID = Argument::get('parent', $argv, null, 'integer');
```
Comment (Quix0r) [inline bin/worker.php]: Maybe rename this to `$parentId` ?

File: include/text.php [hunk 10]
```diff
@@ -652,7 +647,7 @@ function logger($msg, $level = 0) {
 
 	$logline = sprintf("%s@%s\t[%s]:%s:%s:%s\t%s\n",
 			DateTimeFormat::utcNow(DateTimeFormat::ATOM),
-			$process_id,
+			Process::getCode(),
```
Comment (Quix0r) [inline include/text.php]: Maybe rename `Process:getCode()` to `Process:getProcessId()`?

File: mod/worker.php [hunk 11]
```diff
@@ -32,7 +33,9 @@ function worker_init()
 
 	Worker::startProcess();
 
-	logger("Front end worker started: ".getmypid());
+	$process_code = Process::getCode();
```
Comment (Quix0r) [inline mod/worker.php]: Same name part `code` used. Maybe `$processId` is really proper here?

File: mod/worker.php [hunk 12]
```diff
@@ -41,8 +44,8 @@ function worker_init()
 		// But since it doesn't destroy anything, we just try to get more execution time in any way.
 		set_time_limit(0);
 
-		$fields = ['executed' => DateTimeFormat::utcNow(), 'pid' => getmypid(), 'done' => false];
-		$condition =  ['id' => $r[0]["id"], 'pid' => 0];
+		$fields = ['executed' => DateTimeFormat::utcNow(), 'process' => Process::getID(), 'done' => false];
```
Comment (Quix0r) [inline mod/worker.php]: Ah, you have `getID()` and `getCode()` ... Hmm ...

File: src/Core/Worker.php [hunk 13]
```diff
@@ -300,18 +300,16 @@ public static function execute($queue)
 	 * @param array   $argv        Array of values to be passed to the function
 	 * @param boolean $method_call boolean
 	 * @return void
+	 *
+	 * @throws InternalServerErrorException The current process wasn't properly initialized
```
Comment (Quix0r) [inline src/Core/Worker.php]: Is this really thrown somewhere? Cannot find it.

File: src/Core/Worker.php [hunk 14]
```diff
@@ -852,9 +843,9 @@ private static function findWorkerProcesses(&$passing_slow)
 		}
 
 		if ($found) {
-			$condition = "`id` IN (".substr(str_repeat("?, ", count($ids)), 0, -2).") AND `pid` = 0 AND NOT `done`";
+			$condition = "`id` IN (".substr(str_repeat("?, ", count($ids)), 0, -2).") AND `process` = 0 AND NOT `done`";
```
Comment (Quix0r) [inline src/Core/Worker.php]: Maybe name the column `process_id`?

File: bin/daemon.php (hunk 8 above)
Comment (MrPetovan) [reply bin/daemon.php]: No, there's a `process` table with an `id` AND a `pid` field. The `pid` used to be the primary key but it isn't sufficient for multi-host Friendica installations. Hence the add of the more generic `id` primary key field.

File: include/text.php (hunk 10 above)
Comment (MrPetovan) [reply include/text.php]: I agree it's confusing as well, but the `code` is supposed to be an identifier string that would be added to logs. It uses the process id but not only. I don't have a better name for it though, feel free to suggest another, but `getProcessId` is inaccurate.

File: src/Database/DBStructure.php (hunk 5 above)
Comment (annando) [reply src/Database/DBStructure.php]: Nonetheless you can simply drop the two indexes.

File: bin/worker.php (hunk 9 above)
Comment (nupplaphil) [reply bin/worker.php]: Hm .. I'm more a fan of short arguments, but I'm not dogmatic.
I've therefore to check the difference between `parentId` and `parentID` f.e.

File: mod/worker.php (hunk 12 above)
Comment (nupplaphil) [reply mod/worker.php]: `getID()` is for the unique ID and `getCode()` is for the concatenated string :-)

I tried to encapsulate the Process Core and just add some necessary methods

File: src/Core/Worker.php (hunk 13 above)
Comment (nupplaphil) [reply src/Core/Worker.php]: Hm.. You're right. I thought there was a call to a Process method which can caus an Exception. I've to check it again..

File: src/Core/Worker.php (hunk 14 above)
Comment (nupplaphil) [reply src/Core/Worker.php]: You're right.
As far as I can see, it's best practice in the DB-Structure to add the primary key name of the referenced table to the foreign key...

File: src/Database/DBStructure.php (hunk 4 above)
Comment (annando) [reply src/Database/DBStructure.php]: I guess I will add the possibility to run a "pre update" script. There you have to ensure that the data and table structure is prepared for the change. This update script has to add the "id" field (not as primary at this moment), then it has to fill the existing lines. (We mustn't drop any task in the worker queue) After this is done it has to change the primary hey.

I don't think that an automatic update routine could handle this.

File: bin/worker.php (hunk 9 above)
Comment (Quix0r) [reply bin/worker.php]: I mean I like the `camelCase` way. :-)

File: src/Core/Worker.php (hunk 13 above)
Comment (Quix0r) [reply src/Core/Worker.php]: `@throws` should be only annotated when the method in question is throwing it, not when a then called method will throw it. It is a minor documentation mistake, nothing big. :-)

File: bin/worker.php (hunk 9 above)
Comment (MrPetovan) [reply bin/worker.php]: Of course but you can account for argument casing mistakes.

File: src/Core/Worker.php (hunk 13 above)
Comment (nupplaphil) [reply src/Core/Worker.php]: I'm not sure about that. So lets' say I throw an Exception in a inner method. And the outer method itself doesn't handle any exceptions because it's just a "layer" inbetween, the next outer method will possibly get an exception without knowing it and therefore wouldn't create an adequate Exception handling.

The alternative is that I make a catch in and throw this eception again .. But that's uggly I think..

But for this case, we don't need it :-)

ByTheWay: PHPStorm "cries" when there is a possible returning Exception from a called function without a proper Exception handling or a `@throw` in the phpdoc :)

File: bin/worker.php (hunk 9 above)
Comment (nupplaphil) [reply bin/worker.php]: ah now I undertand @Quix0r  .. not `parentID` - it should be `parentId`
I thought I should rename the argument `parent` to `parentId`
=> fully understandable for me and I'll change it

File: include/text.php (hunk 10 above)
Comment (nupplaphil) [reply include/text.php]: What about Process::toString() ?

Comment (MrPetovan) [reply include/text.php]: That's what I initially suggested, but I'm not sure how it would look. Maybe
```php
$processCode = (string) Process::getInstance();
```

Comment (nupplaphil) [reply include/text.php]: I think this doesn't look very intuitive :-( .. Why not:
```php
$processCode = Process::toString();
```
It's shorter, explicit and more intuitive

Comment (MrPetovan) [reply include/text.php]: Fair enough, I'm all for it.

File: bin/worker.php (hunk 9 above)
Comment (Quix0r) [reply bin/worker.php]: @nupplaphil fixed my comment.

File: src/Core/Worker.php (hunk 13 above)
Comment (Quix0r) [reply src/Core/Worker.php]: Okay, since PHP doesn't have `throws` keyword for method signatures.

Comment (nupplaphil) [reply src/Core/Worker.php]: OK, now I found it:

`startProcess()` calls `Process::start()` which throws an Exception which is not handled in `startProcess`. So the `@throws` is on the right place because otherwise the caller of `startProcess()` wouldn't get a hint that it may get an Exception when calling this method.

File: boot.php [hunk 15]
```diff
@@ -838,6 +847,8 @@ function check_addons(App $a)
 /**
  * @brief Used to end the current process, after saving session state.
  * @deprecated
+ *
+ * Use
```
Comment (MrPetovan) [inline boot.php]: Use ?

File: src/Database/DBStructure.php (hunk 4 above)
Comment (nupplaphil) [reply src/Database/DBStructure.php]: That's bad because it seems like this PR can never get automatically deployed :-(

Would it be better to use `pid` for `id` and add a `process_id` for `pid`?
This seems really weird so I won't prefer it.

Another idea:
- set Friendica in maintenance mode
- stop all worker/daemon
- set `workerqueue`.`process_id` to 0
- drop `process`
- recreate `process`

**Comment by nupplaphil** [comment]
So, next step is to fix the unittests ..

File: src/Model/Process.php [hunk 16]
```diff
@@ -292,7 +292,17 @@ public static function stop($kill = false, $id = null)
 	 * @return bool Returns true if the process is inactive
 	 */
 	public static function checkInactive($id) {
-		$entry = dba::selectFirst('process', ['pid'], ['id' => $id, 'parent' => self::$instance->parent]);
+		if (is_null(self::$instance)) {
+			return false;
```
Comment (Quix0r) [inline src/Model/Process.php]: Is this a silent failure? Maybe raise a loud exception here?

File: src/Model/Process.php [hunk 17]
```diff
@@ -305,13 +315,19 @@ public static function checkInactive($id) {
 	 */
 	public static function deleteInactive()
 	{
-		if (is_null(self::$instance) || is_null(self::$instance->parent)) {
+		if (is_null(self::$instance)) {
```
Comment (Quix0r) [inline src/Model/Process.php]: And maybe again a silent failure?

File: src/Model/Process.php [hunk 18]
```diff
@@ -305,13 +315,19 @@ public static function checkInactive($id) {
...
 			return;
 		}
 
+		if (is_null(self::$instance->parent)) {
+			$condition = array("`parent` IN (?, ?)", self::$instance->parent, self::$instance->id);
```
Comment (Quix0r) [inline src/Model/Process.php]: Please use "new" array style `[]` instead.

File: src/Database/DBStructure.php (hunk 4 above)
Comment (annando) [reply src/Database/DBStructure.php]: You could do this in a pre update script. But I don't think that this is a problem for automatic deployment. Why should it?

**Comment by annando** [comment]
After my current work I will try to implement pre update scripts. Then we can proceed with this.

File: src/Database/DBStructure.php (hunk 4 above)
Comment (nupplaphil) [reply src/Database/DBStructure.php]: I thought so because you wrote "I don't think that an automatic update routine could handle this." So I thought therefor it is a blocker for this PR "forever".

Maybe I used the wrong words, sorry

Comment (annando) [reply src/Database/DBStructure.php]: With "automatic update routine" I meant the build in updater in the dbstructure class. This one won't be able to handle it. But a pre-update script should be able. But of course this needs to be tested very well.

**Comment by MrPetovan** [comment]
Sorry for the conflict! It's for the better, I promise.

**Comment by nupplaphil** [comment]
No problem :) I expected your changes and hopefully I'll merge them today.

@rudloff Can you maybe help me with the unittests? When I disable all other unittests, the ProcessTest works. So it seems that there are some preconditions which aren't working with the new ones :-( maybe it's because of the `@runTestsInSeparateProcesses` ?

File: src/Database/DBStructure.php (hunk 4 above)
Comment (nupplaphil) [reply src/Database/DBStructure.php]: Sorry, I thought the pre-update is the update routine. That was the missing link :-)

**Comment by nupplaphil** [comment]
... wow ... thanks @MrPetovan  it seems your PR fixed my unittest failures ..
@Rudloff I'll try it nevertheless with the local annotation .. just to be sure

**Comment by MrPetovan** [comment]
That's how good I am 😎

**Comment by nupplaphil** [comment]
Because I already have to wait until @annando has some time for the pre update function, we don't have to rush with this PR :-)

**Comment by nupplaphil** [comment]
@rudloff at least my old unittests are re-occurring now :-)
But without the `@runInSeperateProcess`, the Singleton-Process wouldn't get initialized properly :-(

**Comment by Rudloff** [comment]
I guess we need way to fix the Smarty error then. Maybe by manually cleaning the template cache between each test?

**Comment by Quix0r** [comment]
Can that rendering be turned off, temporarily? Like you can do with Laravel's Blade templates?

**Comment by Rudloff** [comment]
I had a look a the compiled template files and they all start with a condition like this:
```php
if ($_smarty_tpl->_decodeProperties($_smarty_tpl, array (
  'version' => '3.1.31',
  'unifunc' => 'content_5ad368ea0a79a4_03565794',
  'has_nocache_code' => false,
  'file_dependency' =>
  array (
    '4b4b0827e9042cea499cecd53fffdf3ec896d1ef' =>
    array (
      0 => '/home/pierre/www/friendica/view/templates/shared_content.tpl',
      1 => 1520889983,
      2 => 'file',
    ),
  ),
  'includes' =>
  array (
  ),
),false)) {
```
(I guess it checks if the template needs to be recompiled.)

But when the test runs in a separate process and after another test that uses the same template, `$_smarty_tpl` is null for some reason. But I don't really understand why.

(I tried using [`$force_compile`](https://www.smarty.net/docsv2/en/variable.force.compile.tpl) as a workaround but it didn't work.)

**Comment by nupplaphil** [comment]
@rudloff Thanks for your investigating. I was currently adding the `$force_compile` too.. So I don't have to follow this path :-/

File: tests/src/Util/ArgumentTest.php [hunk 19]
```diff
@@ -23,21 +23,20 @@ function testInvalidParameter() {
 		$this->assertEquals('', $value);
 	}
 
-	/**
-	 * @expectedException Friendica\Network\HTTPException\InternalServerErrorException
```
[the remark this answers is not in the document]
Comment (nupplaphil) [reply tests/src/Util/ArgumentTest.php]: I'll change the name of the test (misleading wording)

File: friendica_test_data.sql [hunk 20]
```diff
@@ -1896,11 +1894,11 @@ CREATE TABLE `workerqueue` (
   `parameter` text,
   `priority` tinyint(3) unsigned NOT NULL DEFAULT '0',
   `created` datetime NOT NULL DEFAULT '0001-01-01 00:00:00',
-  `pid` int(11) NOT NULL DEFAULT '0',
+  `process_id` int(11) NOT NULL DEFAULT '0',
   `executed` datetime NOT NULL DEFAULT '0001-01-01 00:00:00',
   `done` tinyint(1) NOT NULL DEFAULT '0',
   PRIMARY KEY (`id`),
-  KEY `pid` (`pid`),
+  KEY `process_id` (`process_id`),
```
[the remark this answers is not in the document]
Comment (MrPetovan) [reply friendica_test_data.sql]: No, only in case of manual install.

[the remark this answers is not in the document]
Comment (nupplaphil) [reply friendica_test_data.sql]: I guess this will be one of the targets from @annando when he creates a "pre update" function

File: src/Model/Process.php (hunk 16 above)
Comment (nupplaphil) [reply src/Model/Process.php]: No, because maybe it's "just" a process of another node. So I return "false" because I cannot know if the process is inactive

**Comment by Rudloff** [comment]
What I really don't understand is the near empty stack trace:
```
#0 -(978): require_once()
#1 {main}
  thrown in /home/travis/build/friendica/friendica/view/smarty3/compiled/d9640779c31d43468d72999800b0d9eec7dad6fa_0.file.shared_content.tpl.php on line 6
```

Smarty stack traces usually include the PHP file that called the template. For example:
```
#0 /var/www/friendica/vendor/smarty/smarty/libs/sysplugins/smarty_template_compiled.php(107): include()
#1 /var/www/friendica/vendor/smarty/smarty/libs/sysplugins/smarty_template_compiled.php(163): Smarty_Template_Compiled->process(Object(Smarty_Internal_Template))
#2 /var/www/friendica/vendor/smarty/smarty/libs/sysplugins/smarty_internal_template.php(206): Smarty_Template_Compiled->render(Object(Smarty_Internal_Template))
#3 /var/www/friendica/vendor/smarty/smarty/libs/sysplugins/smarty_internal_templatebase.php(216): Smarty_Internal_Template->render(false, 0)
#4 /var/www/friendica/vendor/smarty/smarty/libs/sysplugins/smarty_internal_templatebase.php(107): Smarty_Internal_TemplateBase->_execute(Object(Smarty_Internal_Template), NULL, NULL, NULL, 0)
#5 /var/www/friendica/src/Render/FriendicaSmarty.php(53): Smarty_Inter in /var/www/friendica/view/smarty3/compiled/065f800854856a4ccbf1f076dcab5f95739b9b87_0.file.nav_head.tpl.php on line 6
```

So I don't understand what is happening here. It's like some PHP script in including the remplate directly without using Smarty functions.

File: tests/ApiTest.php [hunk 21]
```diff
@@ -28,7 +29,7 @@ protected function setUp()
 		parent::setUp();
 
 		// Reusable App object
-		$this->app = \Friendica\BaseObject::getApp();
```
[the remark this answers is not in the document]
Comment (nupplaphil) [reply tests/ApiTest.php]: I now know how I can make you happy ;-) :-D

File: src/Database/DBStructure.php (hunk 4 above)
Comment (annando) [reply src/Database/DBStructure.php]: The current update routine can only do a post-update, no pre-update.

File: tests/src/Core/Model/ProcessTest.php [hunk 22]
```diff
@@ -16,13 +15,9 @@
...
 class ProcessTest extends DatabaseTest
 {
-	/**
-	 * @var App
-	 */
-	private $app;
-
 	protected function setUp()
 	{
+		global $a;
```
Comment (MrPetovan) [inline tests/src/Core/Model/ProcessTest.php]: There's no more global `$a` variable and if there is, it's a bug, the app should now be retrieved using `BaseObject::getApp()`.

Comment (nupplaphil) [reply tests/src/Core/Model/ProcessTest.php]: I believe it's a forgotten fragment of the PR before the config-PR, I'll remove it :)

Comment (MrPetovan) [reply tests/src/Core/Model/ProcessTest.php]: The App holds a reference to the SmartyEngine indeed.

**Comment by nupplaphil** [comment]
Back in business with this PR ;-) ..

**Comment by annando** [comment]
@nupplaphil we soon will have a pre update script: https://github.com/friendica/friendica/pull/5425

You have to pay attention to an important fact: The update can be executed from any previous version. So you cannot expect - for example - that the table "workerqueue" even exists or in which structure it exists. Please make sure that the script will be successful in any situation.

**Comment by nupplaphil** [comment]
@annando I tried my best to encapsulate the table altering/renaming inside the `DBStructure.php` for better re-usability.

File: src/Database/DBA.php [hunk 23]
```diff
@@ -643,6 +643,10 @@ public static function exists($table, $condition) {
 
 		$fields = [];
 
+		if (is_null($condition) || empty($condition)) {
```
Comment (MrPetovan) [inline src/Database/DBA.php]: This condition can be simplified to just `empty()`

File: src/Database/DBStructure.php [hunk 24]
```diff
@@ -684,6 +684,141 @@ private static function groupBy($indexname, $fieldnames) {
...
+			return false;
+		}
+
+		if (is_null($columns) || empty($columns)) {
+			return self::existsTable($table);
+		}
+
+		$table = DBA::escape($table);
+
+		foreach ($columns AS $column) {
```
Comment (nupplaphil) [inline src/Database/DBStructure.php]: @annando do you know a better way to check if columns exist in a table? I know that a foreach loop for each column isn't the best way..

**Comment by annando** [comment]
@nupplaphil Your changes are awesome! I hadn't looked too deep into the functionality, but it looks fine. You don't need to apply all changes, you just need to prepare the table so that the structure change can work..

**Comment by nupplaphil** [comment]
@annando I updated my node `friendica.philipp.info` with this branch and I forgot to bump the DB-version.

Nevertheless the output was:
```console
ALTER IGNORE TABLE `process` DROP INDEX `PRIMARY`, ADD `id` int unsigned NOT NULL auto_increment COMMENT 'Global Unique ID', MODIFY `pid` int unsigned COMMENT 'PID of the process', ADD `par
ent` int unsigned COMMENT 'ID of the parent process', ADD `type` varchar(5) NOT NULL COMMENT 'Type of the process', ADD PRIMARY KEY(`id`);
ALTER IGNORE TABLE `thread` ADD `post-type` tinyint unsigned NOT NULL DEFAULT 0 COMMENT 'Post type (personal note, bookmark, ...)', MODIFY `bookmark` boolean COMMENT '';
ALTER IGNORE TABLE `workerqueue` DROP INDEX `pid`, ADD `process_id` int unsigned NOT NULL DEFAULT 0 COMMENT 'Process id of the worker', ADD INDEX `process_id` (`process_id`);
```
It seems like update routine can even rename columns/indices if. Or did I miss something?

**Comment by nupplaphil** [comment]
OK I found bugs on my node in the `worker.php`
- the `parent` column wasn't set right
- there is a mismatch between the PID I see in the table `process` and the PID of `ps -faux`

Nevertheless how does the process table look like under worker load:
```console
MariaDB [friendica]> select * from process;
+-------+------------------------------+---------------------+------+--------+------+
| pid   | command                      | created             | id   | parent | type |
+-------+------------------------------+---------------------+------+--------+------+
| 12915 | /var/www/html/bin/worker.php | 2018-07-21 07:52:58 | 4867 |   NULL | wrk  |
| 14178 | /var/www/html/bin/worker.php | 2018-07-21 07:56:36 | 5334 |   NULL | wrk  |
| 14298 | /var/www/html/bin/worker.php | 2018-07-21 07:56:55 | 5388 |   NULL | wrk  |
| 14304 | /var/www/html/bin/worker.php | 2018-07-21 07:56:57 | 5391 |   NULL | wrk  |
| 14319 | /var/www/html/bin/worker.php | 2018-07-21 07:56:59 | 5398 |   NULL | wrk  |
| 14323 | /var/www/html/bin/worker.php | 2018-07-21 07:56:59 | 5400 |   NULL | wrk  |
| 14364 | /var/www/html/bin/worker.php | 2018-07-21 07:57:04 | 5420 |   NULL | wrk  |
...
| 15205 | /var/www/html/bin/worker.php | 2018-07-21 07:59:09 | 5808 |   NULL | wrk  |
+-------+------------------------------+---------------------+------+--------+------+
69 rows in set (0.000 sec)
```

And that's how the PS looks like:
```console
root     22196  0.0  0.0   1512  1004 ?        Ss   Jul08   0:01  |   |   \_ /bin/sh /cron.sh
root      5135  0.0  0.0   1504     4 ?        S    09:49   0:00  |   |       \_ sleep 10m
root     12336  1.0  0.2 102072 25632 ?        S    09:51   0:03  |   |       \_ php bin/worker.php --cron --parent 3566
root     15219  1.3  0.2 106160 29120 ?        S    09:52   0:04  |   |       \_ php bin/worker.php --cron --parent 4847
root     15915  0.1  0.2 103332 26300 ?        S    09:54   0:00  |   |       \_ php bin/worker.php --cron --parent 4425
root     16132  0.3  0.2 103404 26296 ?        S    09:54   0:00  |   |       \_ php bin/worker.php --cron --parent 3600
root     16382  1.4  0.3 108616 32724 ?        S    09:54   0:02  |   |       \_ php bin/worker.php --cron --parent 4425
root     16464  0.3  0.2 105452 28648 ?        S    09:55   0:00  |   |       \_ php bin/worker.php --cron --parent 4425
root     16667  0.4  0.2 103404 26300 ?        S    09:55   0:00  |   |       \_ php bin/worker.php --cron --parent 3600
root     16675  0.4  0.2 103404 26620 ?        S    09:55   0:00  |   |       \_ php bin/worker.php --cron --parent 4025
root     17034  0.2  0.2 101936 23884 ?        S    09:56   0:00  |   |       \_ php bin/worker.php --cron --parent 3600
root     17041  0.3  0.2 103432 28152 ?        S    09:56   0:00  |   |       \_ php bin/worker.php --cron --parent 4787
root     17232  0.2  0.2 101936 23832 ?        S    09:56   0:00  |   |       \_ php bin/worker.php --cron --parent 4425
root     17260  0.2  0.2  99888 23640 ?        S    09:56   0:00  |   |       \_ php bin/worker.php --cron --parent 4425
root     17274  0.2  0.2  99888 23996 ?        S    09:56   0:00  |   |       \_ php bin/worker.php --cron --parent 3600
root     17294  0.2  0.2 101936 24944 ?        S    09:56   0:00  |   |       \_ php bin/worker.php --cron --parent 3600
root     17304  0.2  0.2  99888 23952 ?        S    09:56   0:00  |   |       \_ php bin/worker.php --cron --parent 3600
....
```

**Comment by nupplaphil** [comment]
Found another issue:
- [ ] hovercard is now throwing errors:
```console
Warning: Cannot modify header information - headers already sent by (output started at /var/www/html/mod/hovercard.php:40) in /var/www/html/index.php on line 512

Warning: Cannot modify header information - headers already sent by (output started at /var/www/html/mod/hovercard.php:40) in /var/www/html/index.php on line 513

Warning: Cannot modify header information - headers already sent by (output started at /var/www/html/mod/hovercard.php:40) in /var/www/html/index.php on line 520

Warning: Cannot modify header information - headers already sent by (output started at /var/www/html/mod/hovercard.php:40) in /var/www/html/index.php on line 521

Warning: Cannot modify header information - headers already sent by (output started at /var/www/html/mod/hovercard.php:40) in /var/www/html/index.php on line 522

Warning: Cannot modify header information - headers already sent by (output started at /var/www/html/mod/hovercard.php:40) in /var/www/html/index.php on line 523
```

File: boot.php [hunk 25]
```diff
@@ -844,7 +853,7 @@ function check_addons(App $a)
  */
 function killme()
 {
-	exit();
+	Process::stop(Process::STOP_EXIT);
```
Comment (MrPetovan) [inline boot.php]: Please don't make use of `killme()`, this is a deprecated function that will be removed in a future version.

Comment (nupplaphil) [reply boot.php]: Puh .. Than I've to check & replace **all** of the `killme()` everywhere in Friendica..
I thought maybe we can make it with another PR

Comment (MrPetovan) [reply boot.php]: I've seen you replaced some `killme()` with `Process::stop()` but then you reverted the change. Is there a specific reason?

Comment (MrPetovan) [reply boot.php]: Fair enough.

**Comment by MrPetovan** [comment]
Can you please rebase against `develop` to rule out the duplicate commits?

**Comment by nupplaphil** [comment]
I'm on the way :-)

**Comment by nupplaphil** [comment]
ah .. damn rebase :-( ..

**Comment by nupplaphil** [comment]
OK I've to leave it for now (unsatisfied!) 
I'll return and try it again :-(

File: boot.php (hunk 25 above)
Comment (Quix0r) [reply boot.php]: Then just bring up another PR with replacement of deprecated function calls with new.

**Comment by Quix0r** [comment]
@nupplaphil `rebase` is always mixing up github because it is bound to commit ids (I think no other choice for them). So just continue. I have reworked my large PRs into smaller and they all got merged.

**Comment by nupplaphil** [comment]
I lost some code during the rebase. So I now separate the prerequisites to other PRs and try to clean up this PR again.

**Comment by nupplaphil** [comment]
At least I do have now a clean PR where I can readd commits ;-)

I'll continue after #5446

**Comment by Quix0r** [comment]
Conflicts found.

**Comment by annando** [comment]
Just let it open. But of course the conflicts could increase.

**Comment by MrPetovan** [comment]
Let's close it until you are in better shape.
